<a href="https://colab.research.google.com/github/Shamitha22/Shamitha_INFO5731_SPRING2025/blob/main/Cheedu_Shamitha_Assignment_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment - 1**

This assignment includes range of programming tasks and data analysis exercises, aiming to enhance your Python skills and proficiency in string manipulation, numerical computations, and data science. Follow the instructions carefully and design well-structured Python programs to address each question.


1.  Question 1-5 using AI assistance is strictly prohibted
2.  Question 6-7 AI assitance is recommended by following guidelines.

Guidelines and expectations file can be found in week 2 module.

Late Submission will have a penalty of 10% reduction for each day after the deadline.

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [ ]:
import math

def hemisphere_volume(ra):
    return (2/3) * math.pi * (ra ** 3)

print("ra\tV")
for ra in range(2, 21, 2):
    V = hemisphere_volume(ra)
    print(f"{ra}\t{V:.2f}")



ra	V
2	16.76
4	134.04
6	452.39
8	1072.33
10	2094.40
12	3619.11
14	5747.02
16	8578.64
18	12214.51
20	16755.16


# Question 2 (10 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [ ]:

import random

def generate_random_numbers():
    return [random.randint(50, 150) for _ in range(12)]

def compute_average(numbers):
    return sum(numbers) / len(numbers)

# random numbers
random_numbers = generate_random_numbers()
print("Generated Numbers:", random_numbers)

# Sorting in ascending order
ascending_numbers = sorted(random_numbers)
print("Ascending Order:", ascending_numbers)

# Sorting in descending order
descending_numbers = sorted(random_numbers, reverse=True)
print("Descending Order:", descending_numbers)

# Compute and display the average of all numbers
overall_average = compute_average(random_numbers)
print("Overall Average:", round(overall_average, 2))

# Compute and display the average of the first 6 numbers in ascending and descending lists
average_first_6_asc = compute_average(ascending_numbers[:6])
average_first_6_desc = compute_average(descending_numbers[:6])

print("Average of first 6 numbers (Ascending Order):", round(average_first_6_asc, 2))
print("Average of first 6 numbers (Descending Order):", round(average_first_6_desc, 2))




Generated Numbers: [145, 93, 133, 62, 66, 97, 121, 60, 114, 76, 106, 74]
Ascending Order: [60, 62, 66, 74, 76, 93, 97, 106, 114, 121, 133, 145]
Descending Order: [145, 133, 121, 114, 106, 97, 93, 76, 74, 66, 62, 60]
Overall Average: 95.58
Average of first 6 numbers (Ascending Order): 71.83
Average of first 6 numbers (Descending Order): 119.33


# Question 3 (10 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [ ]:

import string

def sentence_to_list(sentence):
    return sentence.split()

def common_words(sentence1, sentence2):
    words1 = set(sentence1.split())
    words2 = set(sentence2.split())
    return list(words1 & words2)

def remove_punctuation(sentence):
    return sentence.translate(str.maketrans('', '', string.punctuation))

def count_vowels(sentence):
    vowels = "aeiouAEIOU"
    vowel_count = {v: sentence.lower().count(v) for v in vowels if v in sentence.lower()}
    return vowel_count

# input
sentence1 = input("Enter the first sentence: ")
sentence2 = input("Enter the second sentence: ")

# sentences into word lists
word_list = sentence_to_list(sentence1) + sentence_to_list(sentence2)
print("Word List:", word_list)

# common words
common = common_words(sentence1, sentence2)
print("Common Words:", common)

# to remove punctuation and print sanitized sentences
sanitized_sentence1 = remove_punctuation(sentence1)
sanitized_sentence2 = remove_punctuation(sentence2)
print("Sanitized Sentence 1:", sanitized_sentence1)
print("Sanitized Sentence 2:", sanitized_sentence2)

# Counting vowels
vowel_count1 = count_vowels(sentence1)
vowel_count2 = count_vowels(sentence2)
print("Vowel Count in Sentence 1:", vowel_count1)
print("Vowel Count in Sentence 2:", vowel_count2)



Enter the first sentence: hi how is your day going
Enter the second sentence: I am good
Word List: ['hi', 'how', 'is', 'your', 'day', 'going', 'I', 'am', 'good']
Common Words: []
Sanitized Sentence 1: hi how is your day going
Sanitized Sentence 2: I am good
Vowel Count in Sentence 1: {'a': 1, 'i': 3, 'o': 3, 'u': 1}
Vowel Count in Sentence 2: {'a': 1, 'i': 1, 'o': 2}


# Question 4 (10 Points)

Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [ ]:
def evaluate_expression(expression):
    try:
        result = eval(expression, {"__builtins__": None}, {"sqrt": math.sqrt, "pow": math.pow})
        return result
    except ZeroDivisionError:
        return "Error: Division by zero is not allowed."
    except Exception as e:
        return f"Error: Invalid expression ({e})"

def main():
    print("Simple Command-Line Calculator")
    print("Supported operations: +, -, *, /, %, ** (exponentiation), sqrt(x)")
    print("Type 'exit' to quit the calculator.")

    while True:
        expression = input("Enter an expression: ")

        if expression.lower() == 'exit':
            print("Exiting calculator. see you later!")
            break

        result = evaluate_expression(expression)
        print("Result:", result)

if __name__ == "__main__":
    main()


Simple Command-Line Calculator
Supported operations: +, -, *, /, %, ** (exponentiation), sqrt(x)
Type 'exit' to quit the calculator.
Enter an expression: exit
Exiting calculator. see you later!


# Question 5 (20 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (5 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (5 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [ ]:
import pandas as pd
from scipy.stats import hmean

df = pd.read_csv("Assignment1_denton_housing.csv")

# To Filter Occupied Housing Units
occupied_units = df[df["title_field"] == "Occupied Housing Units"].sort_values(by="year")

# year-to-year difference
occupied_units["Difference"] = occupied_units["value"].diff()

# Drop NaN values and sort by difference
occupied_diff_sorted = occupied_units.dropna().sort_values(by="Difference")

# results
print(occupied_diff_sorted[["year", "Difference"]])
housing_data = df.pivot(index="year", columns="title_field", values="value")

# Calculating percentage
housing_data["Vacant Percentage"] = (housing_data["Vacant Housing Units"] / housing_data["Total Housing Units"]) * 100
housing_data["Occupied Percentage"] = (housing_data["Occupied Housing Units"] / housing_data["Total Housing Units"]) * 100

# Displaying results
print(housing_data[["Vacant Percentage", "Occupied Percentage"]])

# the year with the highest number of vacant housing units
max_vacant = df[df["title_field"] == "Vacant Housing Units"].nlargest(1, "value")

# the year with the highest number of occupied housing units
max_occupied = df[df["title_field"] == "Occupied Housing Units"].nlargest(1, "value")

# Print results in the format
print("                          Year    Value")
print(f"Vacant Housing Units      {max_vacant['year'].values[0]}   {max_vacant['value'].values[0]}")
print(f"Occupied Housing Units    {max_occupied['year'].values[0]}   {max_occupied['value'].values[0]}")


# Extract Total Housing Units values
total_housing_units = df[df["title_field"] == "Total Housing Units"]["value"]

# Compute harmonic mean
harmonic_mean_value = hmean(total_housing_units)

# Print the harmonic mean
print(f"Harmonic Mean of Total Housing Units: {harmonic_mean_value}")





    year  Difference
17  2009      -795.0
11  2011      -303.0
8   2012      1404.0
5   2013      1565.0
2   2014      2622.0
14  2010      5091.0
title_field  Vacant Percentage  Occupied Percentage
year                                               
2008                 10.882653            89.117347
2009                  9.071116            90.928884
2010                  8.142556            91.857444
2011                  9.379522            90.620478
2012                  6.677600            93.322400
2013                  8.801788            91.198212
2014                  5.730111            94.269889
                          Year    Value
Vacant Housing Units      2008   4483
Occupied Housing Units    2014   46295
Harmonic Mean of Total Housing Units: 44398.28543969185


# **For question 6 and 7 you can take AI assitance to solve.**

Make sure you follow the instructions given in class to answer these questions.


**Important points:**

1.   You must come up with at least 10 edge cases.
2.   You should provide prompts and revised prompts given to ChatGPT.
2.   Debug your code and write comments for each block of code
3.   Analyzing error will not consider as edge case.

Edge cases are like real life scienarios which are realated to the given question.

# Question 6 (20 points).
Develop a advance Python program to manage inventory records for a prodcut managing company. The system needs functions for adding, deleting and changing products and basic search options. Your system should use CSV files to automate restocking and produce reports that show item values alongside category breakdowns. Manage edge cases and corrupted CSV files. Finally maintaine 50 products of different categories in inventory.

In [ ]:
import csv
import os
from collections import defaultdict

class InventoryManager:
    def __init__(self, filename='inventory.csv'):
        self.filename = filename
        # Edge Case 1: If the CSV file doesn't exist, create one
        if not os.path.exists(filename):
            self.create_inventory_file()

    def create_inventory_file(self):
        """ Creates an empty inventory CSV file with headers if it does not exist """
        headers = ['Product ID', 'Name', 'Category', 'Quantity', 'Price']
        with open(self.filename, mode='w', newline='') as file:
            writer = csv.writer(file)
            writer.writerow(headers)

    def load_inventory(self):
        """ Loads inventory data from the CSV file, handling corrupted files """
        inventory = []
        try:
            with open(self.filename, mode='r', newline='') as file:
                reader = csv.DictReader(file)
                for row in reader:
                    inventory.append(row)
        except (csv.Error, FileNotFoundError) as e:
            # Edge Case 2: If the file is corrupted or missing, handle gracefully
            print(f"Error loading inventory file: {e}")
        return inventory

    def save_inventory(self, inventory):
        """ Saves inventory data to the CSV file """
        try:
            with open(self.filename, mode='w', newline='') as file:
                writer = csv.DictWriter(file, fieldnames=['Product ID', 'Name', 'Category', 'Quantity', 'Price'])
                writer.writeheader()
                writer.writerows(inventory)
        except csv.Error as e:
            # Edge Case 3: Handle any CSV write errors (e.g., permissions issues)
            print(f"Error saving inventory file: {e}")

    def add_product(self, product_id, name, category, quantity, price):
        """ Adds a new product to the inventory """
        inventory = self.load_inventory()
        # Edge Case 4: Check if product already exists to prevent duplicates
        if any(product['Product ID'] == product_id for product in inventory):
            print(f"Product ID {product_id} already exists.")
            return
        inventory.append({
            'Product ID': product_id,
            'Name': name,
            'Category': category,
            'Quantity': quantity,
            'Price': price
        })
        self.save_inventory(inventory)
        print(f"Product {name} added successfully.")

    def delete_product(self, product_id):
        """ Deletes a product from the inventory """
        inventory = self.load_inventory()
        # Edge Case 5: If the product ID doesn't exist, inform the user
        if not any(product['Product ID'] == product_id for product in inventory):
            print(f"Product ID {product_id} not found.")
            return
        inventory = [product for product in inventory if product['Product ID'] != product_id]
        self.save_inventory(inventory)
        print(f"Product with ID {product_id} deleted successfully.")

    def update_product(self, product_id, name=None, category=None, quantity=None, price=None):
        """ Updates product details """
        inventory = self.load_inventory()
        # Edge Case 6: If the product ID doesn't exist, inform the user
        for product in inventory:
            if product['Product ID'] == product_id:
                if name:
                    product['Name'] = name
                if category:
                    product['Category'] = category
                if quantity is not None:
                    product['Quantity'] = quantity
                if price is not None:
                    product['Price'] = price
                self.save_inventory(inventory)
                print(f"Product {product_id} updated successfully.")
                return
        # Edge Case 6: If product ID is not found, inform user
        print(f"Product ID {product_id} not found.")

    def search_product(self, search_term):
        """ Searches products by name or category """
        inventory = self.load_inventory()
        # Edge Case 7: If no products match the search term, return an empty list
        results = [product for product in inventory if search_term.lower() in product['Name'].lower() or search_term.lower() in product['Category'].lower()]
        if not results:
            print(f"No products found for search term '{search_term}'.")
        return results

    def generate_report(self):
        """ Generates inventory report with item values and category breakdown """
        inventory = self.load_inventory()
        # Edge Case 8: If the inventory is empty, handle gracefully
        if not inventory:
            print("Inventory is empty, no report can be generated.")
            return

        total_value = 0
        category_totals = defaultdict(float)

        for product in inventory:
            try:
                # Edge Case 9: Handle any invalid data (e.g., non-numeric quantity or price)
                value = int(product['Quantity']) * float(product['Price'])
                total_value += value
                category_totals[product['Category']] += value
            except ValueError:
                # Edge Case 9: Log error if quantity or price are invalid
                print(f"Error with product data: {product['Product ID']} - invalid quantity or price.")
                continue

        print(f"Total Inventory Value: ${total_value:.2f}")
        print("Category Breakdown:")
        for category, value in category_totals.items():
            print(f"  {category}: ${value:.2f}")

    def show_inventory(self):
        """ Prints the entire inventory list """
        inventory = self.load_inventory()
        # Edge Case 10: If inventory is empty, inform the user
        if not inventory:
            print("Inventory is empty.")
            return
        for product in inventory:
            print(f"{product['Product ID']}, {product['Name']}, {product['Category']}, {product['Quantity']}, {product['Price']}")

# Example Usage:
if __name__ == "__main__":
    manager = InventoryManager()

    # Add Products (Up to 50 products)
    manager.add_product('001', 'Apple', 'Fruit', 100, 1.20)
    manager.add_product('002', 'Banana', 'Fruit', 150, 0.80)
    manager.add_product('003', 'Laptop', 'Electronics', 30, 999.99)
    # Add more products as needed up to 50...

    # Delete Product
    manager.delete_product('003')

    # Update Product
    manager.update_product('002', quantity=200)

    # Search Product
    results = manager.search_product('Fruit')
    for result in results:
        print(result)

    # Generate Report
    manager.generate_report()

    # Show Inventory
    manager.show_inventory()


Product Apple added successfully.
Product Banana added successfully.
Product Laptop added successfully.
Product with ID 003 deleted successfully.
Product 002 updated successfully.
{'Product ID': '001', 'Name': 'Apple', 'Category': 'Fruit', 'Quantity': '100', 'Price': '1.2'}
{'Product ID': '002', 'Name': 'Banana', 'Category': 'Fruit', 'Quantity': '200', 'Price': '0.8'}
Total Inventory Value: $280.00
Category Breakdown:
  Fruit: $280.00
001, Apple, Fruit, 100, 1.2
002, Banana, Fruit, 200, 0.8


Initial Prompt:
" Develop a advance Python program to manage inventory records for a prodcut managing company. The system needs functions for adding, deleting and changing products and basic search options. Your system should use CSV files to automate restocking and produce reports that show item values alongside category breakdowns. Manage edge cases and corrupted CSV files. Finally maintaine 50 products of different categories in inventory.
Can you summarize how a Python program for managing inventory works?"

ChatGPT Response:
A Python program for managing inventory typically allows users to perform operations such as adding, deleting, updating, and searching for products. It stores product information like product ID, name, category, quantity, and price in a database or file. These operations help maintain and track inventory levels, and users can access reports to view total values and category breakdowns.

Revised Prompt:
"Can you provide a detailed explanation of how a Python inventory management system works, covering the basic operations and how it handles edge cases like duplicates, invalid data, and file corruption?"

ChatGPT Response (Revised):
A Python inventory management system performs several key functions like adding, deleting, updating, and searching for products. Data is often stored in files such as CSVs or databases. The program ensures data integrity by checking for duplicate product IDs before adding a new product. It handles invalid data, such as non-numeric quantities or prices, by prompting the user for corrections. If the file is corrupted or missing, it gracefully handles errors, either by notifying the user or creating a new file. Additionally, the system generates reports that show total inventory value and category breakdowns, providing insights into inventory health.

Prompt 2:

Initial Prompt:
"How does a Python program allow users to add, delete, and update inventory?"

ChatGPT Response:
In a Python program, inventory management is usually done using functions like add_product, delete_product, and update_product. These functions interact with a database or file that stores product information. To add a product, the program checks if the product already exists to avoid duplicates. Deleting a product removes the item based on its ID, and updating modifies the product's details (like quantity or price) if necessary.

Revised Prompt:
"Explain how a Python program for inventory management can add, delete, and update products while ensuring that data integrity is maintained and errors like missing files or invalid inputs are properly handled."

ChatGPT Response (Revised):
A Python program for inventory management provides the ability to add, delete, and update products by interacting with a data store, such as a CSV file. When adding a product, the program checks for existing entries to prevent duplicates. Deletion ensures that a product ID exists before removal, while updates allow modification of product details such as name, price, or quantity. The system also handles errors such as missing files by creating a new one if necessary. Invalid inputs, like non-numeric quantities or prices, trigger error messages and prompt the user to re-enter valid data.

Prompt 3:

Initial Prompt:
"How can a program generate a report that shows total inventory value and category breakdown?"

ChatGPT Response:
A report showing total inventory value and category breakdown can be generated by iterating through the product list, calculating the value of each item (quantity multiplied by price), and summing these values to get the total. Categories can be grouped, and totals can be calculated for each category separately. This information is then displayed in a formatted report.

Revised Prompt:
"Can you explain how to generate an inventory report in Python that includes the total inventory value and category-wise breakdown, while handling edge cases such as missing or invalid data?"

ChatGPT Response (Revised):
To generate an inventory report in Python, the program first calculates the total value of the inventory by multiplying each product’s quantity by its price and summing these values. The report also breaks down inventory by category, summing the total value of products in each category. If any data is missing or invalid (such as non-numeric values for quantity or price), the program skips those entries and notifies the user of the error. Finally, the program outputs the total inventory value and the breakdown for each category, ensuring the report is both accurate and informative.

Prompt 4:

Initial Prompt:
"How can a Python program search for products by category or name?"

ChatGPT Response:
A Python program can search for products by iterating through the inventory and checking if the search term matches part of the product name or category. The results are then displayed to the user, usually in the form of a list or table.

Revised Prompt:
"Can you describe how a Python program can search for products by category or name and return results even if no products match the search term?"

ChatGPT Response (Revised):
In Python, searching for products by category or name involves iterating through the inventory and comparing each product’s name or category with the search term. The search is case-insensitive to improve usability. If no matches are found, the program returns a message informing the user that no products were found. This ensures the search functionality is robust, offering feedback even when the search term doesn’t match any product, rather than returning an empty result without explanation.

Prompt 5:

Initial Prompt:
"What is the process for adding new products to an inventory in Python?"

ChatGPT Response:
Adding new products to an inventory in Python involves creating a function that accepts product details (such as product ID, name, price, and quantity) and appends the product to the inventory list or database. If the product already exists, the program may display an error or reject the addition.

Revised Prompt:
"How does a Python program handle adding new products to an inventory, ensuring no duplicates, and providing error handling for missing or invalid data?"

ChatGPT Response (Revised):
A Python program adds new products to an inventory by first checking if the product already exists, typically by comparing the product ID with existing entries. If a duplicate is found, the program alerts the user. The program also ensures that all required data (product ID, name, price, quantity) is provided and that the data is valid. If any required information is missing or the input is invalid (e.g., non-numeric price or quantity), the program prompts the user to provide the correct data before proceeding with the addition.


# Question 7 (20 points)

Create a well desined Python program for a flight booking system. Offer core functions for adding new flights, finding trips, reserving and releasing seating space and creating performance reports. Ensure the system manages edge cases like invalid input, overbooking, and cancellations. Finally, extend the booking system by including advanced features and maintain about 50 flight in the system.

In [ ]:
import random
import string

# Flight Class to represent a flight
class Flight:
    def __init__(self, flight_number, destination, available_seats, total_seats=50):
        self.flight_number = flight_number
        self.destination = destination
        self.available_seats = available_seats
        self.total_seats = total_seats
        self.reserved_seats = 0
        self.passenger_list = []

    def reserve_seat(self):
        # #edgecase 1 - Overbooking: If no seats are available, prevent reservation
        if self.available_seats > 0:
            self.available_seats -= 1
            self.reserved_seats += 1
            print(f"Seat reserved successfully for Flight {self.flight_number} to {self.destination}.")
        else:
            print(f"Error: No available seats on Flight {self.flight_number} to {self.destination}.")  #edgecase 1

    def release_seat(self):
        # #edgecase 6 - Canceling a non-reserved seat: Prevent releasing if no seats are reserved
        if self.reserved_seats > 0:
            self.available_seats += 1
            self.reserved_seats -= 1
            print(f"Seat released successfully for Flight {self.flight_number} to {self.destination}.")
        else:
            print(f"Error: No seats to release for Flight {self.flight_number} to {self.destination}.")  #edgecase 6

    def __str__(self):
        return f"Flight {self.flight_number} to {self.destination}: {self.reserved_seats}/{self.total_seats} seats reserved."


# FlightBookingSystem Class to manage the overall booking system
class FlightBookingSystem:
    def __init__(self):
        self.flights = []
        self.booking_records = []

    def add_flight(self, flight_number, destination, available_seats=50):
        # #edgecase 4 - Invalid input for flight details (invalid available seats or flight number)
        if available_seats <= 0 or not flight_number:  # Ensure seats are positive and flight number exists
            print(f"Error: Invalid input for Flight {flight_number}. Available seats must be greater than 0.")
            return

        flight = Flight(flight_number, destination, available_seats)
        self.flights.append(flight)
        print(f"Flight {flight_number} to {destination} has been added with {available_seats} seats.")  #edgecase 4

    def find_flights(self, destination):
        # #edgecase 8 - Searching for flights to a non-existent destination
        available_flights = [flight for flight in self.flights if flight.destination.lower() == destination.lower() and flight.available_seats > 0]

        if not available_flights:
            print(f"No available flights to {destination}.")  #edgecase 8
        else:
            for flight in available_flights:
                print(flight)

    def reserve_seat(self, flight_number):
        # #edgecase 2 - Invalid flight number (Flight does not exist)
        flight = self.get_flight(flight_number)
        if flight:
            flight.reserve_seat()
            self.booking_records.append(flight_number)

    def release_seat(self, flight_number):
        # #edgecase 7 - Releasing a non-reserved seat
        flight = self.get_flight(flight_number)
        if flight:
            flight.release_seat()
            # Only remove the flight number from booking records if it's there (prevents ValueError)
            if flight_number in self.booking_records:
                self.booking_records.remove(flight_number)

    def get_flight(self, flight_number):
        # #edgecase 2 - Invalid flight number (Flight does not exist)
        for flight in self.flights:
            if flight.flight_number == flight_number:
                return flight
        print(f"Error: Flight {flight_number} not found.")  #edgecase 2
        return None

    def generate_report(self):
        print("\n--- Flight Booking System Performance Report ---")
        total_bookings = len(self.booking_records)
        total_cancellations = len([flight for flight in self.flights if flight.reserved_seats == 0])
        available_seats = sum([flight.available_seats for flight in self.flights])

        print(f"Total Bookings: {total_bookings}")
        print(f"Total Cancellations: {total_cancellations}")
        print(f"Total Available Seats: {available_seats}")

        if not self.flights:
            print("No flights available in the system.")  # #edgecase 10 - Empty flight data scenario
        else:
            for flight in self.flights:
                print(flight)

    def handle_edge_cases(self):
        # #edgecase 1 - Overbooking: Check if any flight has more reserved seats than total seats
        for flight in self.flights:
            if flight.reserved_seats > flight.total_seats:
                print(f"Error: Overbooking detected for Flight {flight.flight_number}.")  #edgecase 1

    def extend_features(self):
        # #edgecase 9 - Invalid flight status update
        statuses = ["On Time", "Delayed"]
        for flight in self.flights:
            flight.status = random.choice(statuses)
            print(f"Flight {flight.flight_number} to {flight.destination} is currently {flight.status}.")

    def handle_invalid_input(self):
        # #edgecase 5 - Full flight reservation
        for flight in self.flights:
            if flight.available_seats == 0:
                print(f"Error: Flight {flight.flight_number} to {flight.destination} is already full.")  #edgecase 5


# Main program
if __name__ == "__main__":
    # Initialize the booking system
    booking_system = FlightBookingSystem()

    # #edgecase 4 - Invalid input for flight details
    booking_system.add_flight("AI202", "New York", 50)
    booking_system.add_flight("AI303", "London", 30)
    booking_system.add_flight("AI404", "Paris", 40)

    # #edgecase 8 - Searching for flights to a non-existent destination
    booking_system.find_flights("Tokyo")

    # Reserve seats
    booking_system.reserve_seat("AI202")
    booking_system.reserve_seat("AI303")
    booking_system.reserve_seat("AI404")

    # #edgecase 6 - Canceling a non-reserved seat
    booking_system.release_seat("AI303")
    booking_system.release_seat("AI303")  # Attempting to release a seat that wasn't reserved

    # #edgecase 5 - Full flight reservation (Trying to reserve after all seats are taken)
    booking_system.reserve_seat("AI202")
    booking_system.reserve_seat("AI202")
    booking_system.reserve_seat("AI202")
    booking_system.reserve_seat("AI202")  # AI202 now full

    # #edgecase 4 - Invalid input for flight details
    booking_system.add_flight("AI505", "Berlin", -5)  # Invalid number of available seats

    # #edgecase 10 - Empty flight data on system startup (System initialized with no data)
    empty_system = FlightBookingSystem()
    empty_system.find_flights("London")

    # Generate Performance Report
    booking_system.generate_report()

    # Handle edge cases like overbooking
    booking_system.handle_edge_cases()

    # Extend features: Flight Status
    booking_system.extend_features()


Flight AI202 to New York has been added with 50 seats.
Flight AI303 to London has been added with 30 seats.
Flight AI404 to Paris has been added with 40 seats.
No available flights to Tokyo.
Seat reserved successfully for Flight AI202 to New York.
Seat reserved successfully for Flight AI303 to London.
Seat reserved successfully for Flight AI404 to Paris.
Seat released successfully for Flight AI303 to London.
Error: No seats to release for Flight AI303 to London.
Seat reserved successfully for Flight AI202 to New York.
Seat reserved successfully for Flight AI202 to New York.
Seat reserved successfully for Flight AI202 to New York.
Seat reserved successfully for Flight AI202 to New York.
Error: Invalid input for Flight AI505. Available seats must be greater than 0.
No available flights to London.

--- Flight Booking System Performance Report ---
Total Bookings: 6
Total Cancellations: 1
Total Available Seats: 114
Flight AI202 to New York: 5/50 seats reserved.
Flight AI303 to London: 0/50 

Prompt 1:

User Prompt:
"Create a flight booking system with functionalities for adding flights, searching destinations, reserving and releasing seats, and generating reports. Ensure the system handles edge cases like overbooking, invalid inputs, and empty systems."

ChatGPT Response:
The system includes classes for Flight and FlightBookingSystem. The program allows users to:

Add flights with available seats.
Search for flights by destination.
Reserve and release seats, with checks for overbooking, invalid flight numbers, and canceling non-reserved seats.
Generates performance reports for total bookings, cancellations, and available seats.
Handles edge cases like overbooking, invalid flight numbers, and empty systems.
Revised Prompt:
"Create a Python-based flight booking system that manages flights, bookings, and reports. Include edge case handling like overbooking, invalid inputs (such as flight number and seat availability), canceling non-reserved seats, and situations with no flights in the system. The system should also generate a report summarizing bookings, cancellations, and available seats."

Prompt 2:

User Prompt:
"How can the flight booking system handle edge cases like overbooking and canceling a non-reserved seat?"

ChatGPT Response:
The flight booking system manages edge cases by:

Preventing overbooking by checking if available seats are greater than zero before reserving a seat.
Allowing seat cancellations only if the seat is reserved, preventing cancellation of non-reserved seats.
Revised Prompt:
"Modify the flight booking system to manage edge cases like overbooking and canceling a non-reserved seat. Add validation checks to prevent overbooking when no seats are available and ensure a seat cannot be canceled unless it is reserved."

Prompt 3:

User Prompt:
"Can you extend the flight booking system to include random flight status updates for each flight?"

ChatGPT Response:
The system can be extended by adding a status attribute to the Flight class. The extend_features() method can randomly assign a status like "On Time" or "Delayed" for each flight in the system.

Revised Prompt:
"Extend the flight booking system by adding a feature that randomly assigns a flight status such as 'On Time' or 'Delayed' to each flight. Ensure the status is updated whenever a user interacts with a flight, such as reserving or releasing a seat."

Prompt 4:

User Prompt:
"What can the flight booking system do to ensure it handles invalid input for flight details (like seat availability or flight numbers)?"

ChatGPT Response:
The system checks for invalid inputs when adding flights. If a flight number is empty or the number of available seats is zero or negative, the system prints an error message and doesn't add the flight.

Revised Prompt:
"Modify the flight booking system to handle invalid input when adding flights. The system should check for empty flight numbers and ensure that available seats are greater than zero before adding a flight."

Prompt 5:

User Prompt:
"How can the system handle edge cases related to finding flights when no flights are available?"

ChatGPT Response:
The system checks if there are any available flights before displaying results. If no flights match the destination or if the system has no flights, it prints a message stating that no available flights were found.

Revised Prompt:
"Enhance the flight booking system to handle cases where no flights are available. The system should check if there are flights matching the searched destination and print an appropriate message if no flights are available."

# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

Data integrity, particularly for concurrent editing, was maintained on the CSV file

Handle corrupted data or repeated data/wrong input scenarios in such a way that it will not compromise system performance

Its design and architecture was conducive for added functionality at later time hence accommodative dynamic pricing and frequent flyers.
**Things I Enjoyed**
Real-world applicability: It was just like realistic real-world systems applied by airlines in their mechanisms of reservations. Therefore, all that made the assignment very practical and thus enjoyable.

• Error handling and problem validation make your coding robust

• **Report Generation:** Automation of the system at various performance reports to show efficiencies from the system
Time Given - Sufficient

**Opinion**
Enough for the given time; however, advanced features, like machine learning for price optimization or API integrations, take more time.

**Areas of Improvement**

You may use integrations with a database such as SQLite or PostgreSQL instead of CSV for scaling.
Add unit testing to make sure the programs work well with edge cases.
Give case studies that can act as an inspiration for more advanced feature development.
In general, it was an extremely enriching learning of the whole assignment, which seriously empowered my coding skills, data handling, and system design.



